File: src/types/effects.ts
/**
 * Clip effects.
 *
 * An effect instance is deliberately dumb — a kind plus a bag of numbers — so
 * the document format never has to change when a new filter is added. All the
 * knowledge lives in the descriptor registry below: parameter ranges, live CSS
 * mapping, and the label the UI shows. Adding a filter is one entry here; the
 * inspector, the preview and the render plan all follow automatically.
 *
 * The ffmpeg mapping deliberately lives elsewhere (`src-tauri/src/engine`, with
 * a TypeScript twin in `src/lib/effectChain.ts`) because the render engine owns
 * it, not the UI.
 */

export type EffectKind =
  | 'brightness'
  | 'contrast'
  | 'saturation'
  | 'blur'
  | 'hue'
  | 'grayscale'
  | 'invert'
  | 'rgbsplit'
  | 'motionblur';

export interface Effect {
  id: string;
  kind: EffectKind;
  enabled: boolean;
  /** Keyed by the descriptor's parameter `key`. */
  params: Record<string, number>;
}

export interface EffectParamSpec {
  key: string;
  label: string;
  min: number;
  max: number;
  step: number;
  /** The value that means "no change" — used by the reset control. */
  neutral: number;
  /** Value applied when the effect is first added, so it is visibly doing something. */
  initial: number;
  /**
   * Factor between the number a user types and the one that is stored — 100 for
   * a parameter shown as a percentage, so `130` means 1.30.
   */
  inputScale?: number;
  format(value: number): string;
}

export interface EffectDescriptor {
  kind: EffectKind;
  label: string;
  hint: string;
  params: EffectParamSpec[];
  /**
   * Live preview fragment, e.g. `saturate(1.3)`.
   *
   * `unit` converts **project pixels into the pixels the viewer is drawing in**
   * — the stage is a scaled-down copy of the frame, and a filter is not scaled
   * by the transform that positions the layer. A length that ignores it looks
   * right on a maximised window and two or three times too strong on a small
   * one, while the export does something else again. Only lengths use it.
   *
   * `null` when the parameters are neutral, and on the effects CSS has no
   * function for at all — those answer through {@link EffectDescriptor.svg}.
   */
  css(params: Record<string, number>, unit: number): string | null;
  /**
   * The SVG filter this effect needs, for the ones the CSS function list
   * cannot express — a channel offset, a blur along one axis.
   *
   * Present exactly on those effects, and `null` when their parameters are
   * neutral. The chain builder puts a `url(#…)` reference in the CSS string
   * where the effect sits, so the stack order survives.
   */
  svg?(params: Record<string, number>, unit: number, id: string): SvgFilterSpec | null;
}

/**
 * A filter the viewer has to draw as SVG rather than as a CSS function.
 *
 * Every length is already in the viewer's own pixels: whoever builds one has
 * applied `unit`, so the renderer places the numbers verbatim.
 */
export type SvgFilterSpec =
  | {
      id: string;
      kind: 'rgbsplit';
      /** How far the red channel moves; blue moves the opposite way. */
      dx: number;
      dy: number;
    }
  | {
      id: string;
      kind: 'motionblur';
      /** Gaussian standard deviation per axis. */
      x: number;
      y: number;
    };

const percent = (value: number) => `${Math.round(value * 100)} %`;
const signedPercent = (value: number) =>
  `${value > 0 ? '+' : ''}${Math.round(value * 100)} %`;

export const EFFECTS: EffectDescriptor[] = [
  {
    kind: 'brightness',
    label: 'Luminosité',
    hint: 'Éclaircit ou assombrit l’image',
    params: [
      {
        key: 'amount',
        label: 'Intensité',
        min: -1,
        max: 1,
        step: 0.01,
        neutral: 0,
        initial: 0.15,
        inputScale: 100,
        format: signedPercent,
      },
    ],
    // CSS brightness is a multiplier around 1; the stored value is an offset,
    // which matches ffmpeg's `eq=brightness` range of -1 → 1.
    css: (params) => `brightness(${(1 + (params.amount ?? 0)).toFixed(3)})`,
  },
  {
    kind: 'contrast',
    label: 'Contraste',
    hint: 'Écarte ou resserre les tons',
    params: [
      {
        key: 'amount',
        label: 'Intensité',
        min: 0,
        max: 2.5,
        step: 0.01,
        neutral: 1,
        initial: 1.2,
        inputScale: 100,
        format: percent,
      },
    ],
    css: (params) => `contrast(${(params.amount ?? 1).toFixed(3)})`,
  },
  {
    kind: 'saturation',
    label: 'Saturation',
    hint: 'Densité des couleurs',
    params: [
      {
        key: 'amount',
        label: 'Intensité',
        min: 0,
        max: 3,
        step: 0.01,
        neutral: 1,
        initial: 1.35,
        inputScale: 100,
        format: percent,
      },
    ],
    css: (params) => `saturate(${(params.amount ?? 1).toFixed(3)})`,
  },
  {
    kind: 'blur',
    label: 'Flou',
    hint: 'Flou gaussien',
    params: [
      {
        key: 'radius',
        label: 'Rayon',
        min: 0,
        max: 40,
        step: 0.5,
        neutral: 0,
        initial: 6,
        format: (value) => `${value.toFixed(1)} px`,
      },
    ],
    /*
     * Halved, and scaled to the viewer.
     *
     * CSS `blur(n)` *is* a Gaussian of standard deviation `n` — the spec says
     * so, and `types/backdrop` has said so here for as long as backdrops have
     * existed. The encoder is handed `radius / 2`, so passing the raw radius to
     * the viewer made the preview exactly twice as blurred as the file it was
     * previewing. The export is left alone and the viewer is corrected to it.
     */
    css: (params, unit) =>
      (params.radius ?? 0) > 0
        ? `blur(${(((params.radius ?? 0) / 2) * unit).toFixed(2)}px)`
        : null,
  },
  {
    kind: 'hue',
    label: 'Teinte',
    hint: 'Rotation de la roue chromatique',
    params: [
      {
        key: 'angle',
        label: 'Angle',
        min: -180,
        max: 180,
        step: 1,
        neutral: 0,
        initial: 30,
        format: (value) => `${value > 0 ? '+' : ''}${Math.round(value)}°`,
      },
    ],
    css: (params) => `hue-rotate(${Math.round(params.angle ?? 0)}deg)`,
  },
  {
    kind: 'grayscale',
    label: 'Noir & blanc',
    hint: 'Désaturation totale progressive',
    params: [
      {
        key: 'amount',
        label: 'Dosage',
        min: 0,
        max: 1,
        step: 0.01,
        neutral: 0,
        initial: 1,
        inputScale: 100,
        format: percent,
      },
    ],
    css: (params) => `grayscale(${(params.amount ?? 0).toFixed(3)})`,
  },
  {
    kind: 'invert',
    label: 'Négatif',
    hint: 'Inverse les couleurs — à pleine dose, la photo en négatif',
    params: [
      {
        key: 'amount',
        label: 'Dosage',
        min: 0,
        max: 1,
        step: 0.01,
        neutral: 0,
        // Full, because a half-inverted frame is flat grey and reads as a bug.
        // Anyone who wants the in-between can drag towards it.
        initial: 1,
        inputScale: 100,
        format: percent,
      },
    ],
    css: (params) => `invert(${(params.amount ?? 0).toFixed(3)})`,
  },
  {
    /*
     * Chromatic aberration.
     *
     * Red goes one way, blue the other, green stays put — which is what a cheap
     * lens does at the edge of its coverage, and what a tape head does when it
     * is losing tracking. On an impact it reads as the frame being struck.
     *
     * Neither CSS nor the canvas has a function for moving one channel, so this
     * is drawn as an SVG filter. The encoder's `rgbashift` moves the same
     * channels by the same signed pixel counts in the same directions — both
     * checked against the binary rather than assumed — so the two are one
     * operation described twice, not two that resemble each other.
     */
    kind: 'rgbsplit',
    label: 'Aberration chromatique',
    hint: 'Décale les canaux rouge et bleu — l’écart d’objectif d’un impact',
    params: [
      {
        key: 'amount',
        label: 'Écart',
        min: 0,
        // `rgbashift` takes a signed byte's worth of pixels; nothing legible
        // needs a tenth of that, and the cap keeps a slider drag sane.
        max: 60,
        // Whole project pixels, because the encoder's option is an integer. A
        // step the export would round away is a step that lies.
        step: 1,
        neutral: 0,
        initial: 10,
        format: (value) => `${Math.round(value)} px`,
      },
      {
        key: 'angle',
        label: 'Direction',
        min: -180,
        max: 180,
        step: 1,
        neutral: 0,
        initial: 0,
        format: (value) => `${value > 0 ? '+' : ''}${Math.round(value)}°`,
      },
    ],
    css: () => null,
    svg: (params, unit, id) => {
      const amount = params.amount ?? 0;
      if (amount <= 0) return null;
      const radians = ((params.angle ?? 0) * Math.PI) / 180;
      return {
        id,
        kind: 'rgbsplit',
        dx: amount * Math.cos(radians) * unit,
        dy: amount * Math.sin(radians) * unit,
      };
    },
  },
  {
    /*
     * Directional blur — the smear that keeps a hard cut from reading as a jump.
     *
     * Gaussian, and **elliptical rather than rotated**: the deviation is split
     * across the two axes, so 0° and 90° are exact and everything between is an
     * axis-aligned ellipse rather than a true streak along the angle. That is a
     * real limitation, and it is the honest one — a rotated Gaussian means
     * turning, blurring and turning back, and the viewer and the encoder would
     * then have to agree on the resampling of both turns to the pixel. An
     * ellipse the two compute identically beats a streak they compute
     * differently.
     *
     * SVG's `stdDeviation="x y"` and ffmpeg's `gblur=sigma=x:sigmaV=y` are the
     * same two numbers in the same units — the equivalence `types/backdrop`
     * already rests on, now stated per axis.
     */
    kind: 'motionblur',
    label: 'Flou de mouvement',
    hint: 'Flou gaussien sur un seul axe — lie les images d’une coupe rapide',
    params: [
      {
        key: 'amount',
        label: 'Intensité',
        min: 0,
        max: 60,
        step: 0.5,
        neutral: 0,
        initial: 14,
        format: (value) => `${value.toFixed(1)} px`,
      },
      {
        key: 'angle',
        label: 'Direction',
        min: -180,
        max: 180,
        step: 1,
        neutral: 0,
        initial: 0,
        format: (value) => `${value > 0 ? '+' : ''}${Math.round(value)}°`,
      },
    ],
    css: () => null,
    svg: (params, unit, id) => {
      const amount = params.amount ?? 0;
      if (amount <= 0) return null;
      const radians = ((params.angle ?? 0) * Math.PI) / 180;
      // Absolute: a blur has an axis, not a direction. 180° is 0°.
      return {
        id,
        kind: 'motionblur',
        x: amount * Math.abs(Math.cos(radians)) * unit,
        y: amount * Math.abs(Math.sin(radians)) * unit,
      };
    },
  },
];

export const effectDescriptor = (kind: EffectKind): EffectDescriptor =>
  EFFECTS.find((item) => item.kind === kind) ?? (EFFECTS[0] as EffectDescriptor);

/** Parameter bag for a freshly added effect. */
export function initialParams(kind: EffectKind): Record<string, number> {
  const out: Record<string, number> = {};
  for (const spec of effectDescriptor(kind).params) out[spec.key] = spec.initial;
  return out;
}

export function neutralParams(kind: EffectKind): Record<string, number> {
  const out: Record<string, number> = {};
  for (const spec of effectDescriptor(kind).params) out[spec.key] = spec.neutral;
  return out;
}

/** True when every parameter sits at its no-op value. */
export function isNeutral(effect: Effect): boolean {
  return effectDescriptor(effect.kind).params.every(
    (spec) => Math.abs((effect.params[spec.key] ?? spec.neutral) - spec.neutral) < 1e-6,
  );
}

/** Everything a viewer needs in order to draw one clip's effect stack. */
export interface FilterChain {
  /** Value for CSS `filter`, or `undefined` when nothing applies. */
  css: string | undefined;
  /** The filters the `url(#…)` references in `css` point at, in that order. */
  svg: SvgFilterSpec[];
}

export const EMPTY_CHAIN: FilterChain = { css: undefined, svg: [] };

/**
 * The whole stack, as one CSS `filter` value plus the SVG it leans on.
 *
 * Order is preserved, and that is why the two halves come back together rather
 * than as "the CSS ones" and "the SVG ones": a stack of blur, split, saturation
 * composes in that order, and a builder that gathered each kind into its own
 * pass would quietly reorder it. An SVG-backed effect contributes a `url(#…)`
 * **at its own position** in the chain, so what is composed is what the
 * inspector lists and what the encoder builds.
 *
 * `unit` turns project pixels into the pixels the caller is drawing in — the
 * stage's scale for the viewer, and 1 for the export bake, which already works
 * at the project's own resolution.
 *
 * `scope` makes the generated ids unique. It is the clip id everywhere, so two
 * clips carrying the same effect never fight over one `<filter>`.
 */
export function filterChainFor(
  effects: Effect[] | undefined,
  unit: number,
  scope: string,
): FilterChain {
  if (!effects || effects.length === 0) return EMPTY_CHAIN;

  const parts: string[] = [];
  const svg: SvgFilterSpec[] = [];

  effects.forEach((effect, index) => {
    if (!effect.enabled) return;
    const descriptor = effectDescriptor(effect.kind);

    if (descriptor.svg) {
      const spec = descriptor.svg(effect.params, unit, `vgfx-${scope}-${index}`);
      if (spec) {
        svg.push(spec);
        parts.push(`url(#${spec.id})`);
      }
      return;
    }

    const css = descriptor.css(effect.params, unit);
    if (css) parts.push(css);
  });

  return parts.length > 0 ? { css: parts.join(' '), svg } : EMPTY_CHAIN;
}

File: src-tauri/src/engine/effects.rs
//! Effect → ffmpeg mapping.
//!
//! This is the authoritative translation from the editor's filter stack to a
//! filtergraph. `src/lib/effectChain.ts` carries a twin used by the browser
//! build; the inspector reports which one produced the chain it is showing, so
//! a drift between the two is visible rather than silent.

use serde::Serialize;

use crate::model::Effect;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EffectChain {
    /// One fragment per enabled, non-neutral effect, in stack order.
    pub filters: Vec<String>,
    /// Those fragments joined into a filtergraph chain.
    pub chain: String,
    pub engine: String,
}

/// A partial colour inversion, as a lookup table.
///
/// CSS defines `invert(a)` as the straight line `out = in·(1−2a) + a` over the
/// 0 → 1 range — at `a = 1` that is `1 − in`, the negative, and at `a = 0.5` it
/// is flat mid-grey. `lutrgb` evaluates the identical line over 0 → `maxval`,
/// per channel, in RGB. So the preview's filter and the encoder's are not an
/// approximation of one another: they are the same function.
///
/// `lutrgb` rather than `negate` because `negate` has no dosage, and rather
/// than `eq=contrast=-1` because ffmpeg's `eq` works on luma alone — that
/// inverts the brightness of a frame while leaving its colours where they were,
/// which is not what anyone means by a negative.
///
/// Shared with the TypeScript twin in `src/lib/effectChain.ts`, character for
/// character. The test below is the contract.
pub fn invert_filter(amount: f64) -> String {
    let ramp = format!("val*({:.3})+({amount:.3})*maxval", 1.0 - 2.0 * amount);
    format!("lutrgb=r={ramp}:g={ramp}:b={ramp}")
}

/// The channel offsets a chromatic split asks for.
///
/// `rgbashift` counts in whole pixels, so the amount is rounded here rather
/// than left to the option parser — the viewer rounds to the same integers, and
/// a value only one of them rounds is a value the two disagree about.
///
/// The signs were checked against the binary: `rh=+10` moves red ten pixels to
/// the **right**, `rv=+8` moves it eight **down** — exactly what SVG's
/// `feOffset dx dy` does in the viewer. Blue takes the opposite of both.
///
/// `edge=smear` holds the border pixel out to the edge rather than leaving the
/// uncovered strip transparent, which is the closest the encoder gets to a
/// filter region that reaches past the layer, and the one that does not put a
/// coloured band down the side of the picture.
///
/// `None` when the offsets round to nothing — a filter that shifts by zero
/// costs a pass and changes no pixel.
///
/// Shared with the TypeScript twin in `src/lib/effectChain.ts`, character for
/// character. The test below is the contract.
pub fn rgb_split_filter(amount: f64, angle: f64) -> Option<String> {
    let radians = angle.to_radians();
    let dx = (amount * radians.cos()).round() as i64;
    let dy = (amount * radians.sin()).round() as i64;
    if dx == 0 && dy == 0 {
        return None;
    }
    Some(format!(
        "rgbashift=rh={dx}:rv={dy}:bh={}:bv={}:edge=smear",
        -dx, -dy
    ))
}

/// A Gaussian blurred along one axis.
///
/// `sigma` is horizontal and `sigmaV` vertical, both standard deviations in
/// pixels — the same two numbers SVG's `stdDeviation="x y"` takes. Passing
/// `sigmaV` explicitly is not optional: left out it defaults to -1, meaning
/// "the same as sigma", and a directional blur would come back isotropic.
///
/// Shared with the TypeScript twin, character for character.
pub fn motion_blur_filter(amount: f64, angle: f64) -> Option<String> {
    let radians = angle.to_radians();
    let x = amount * radians.cos().abs();
    let y = amount * radians.sin().abs();
    if x < 0.05 && y < 0.05 {
        return None;
    }
    Some(format!("gblur=sigma={x:.2}:sigmaV={y:.2}:steps=2"))
}

/// Neutral values are skipped: a filter that changes nothing costs a pass.
pub fn effect_filters(effects: &[Effect]) -> Vec<String> {
    let mut filters = Vec::new();

    for effect in effects {
        if !effect.enabled {
            continue;
        }

        match effect.kind.as_str() {
            "brightness" => {
                let amount = effect.param("amount", 0.0);
                if amount != 0.0 {
                    filters.push(format!("eq=brightness={amount:.3}"));
                }
            }
            "contrast" => {
                let amount = effect.param("amount", 1.0);
                if amount != 1.0 {
                    filters.push(format!("eq=contrast={amount:.3}"));
                }
            }
            "saturation" => {
                let amount = effect.param("amount", 1.0);
                if amount != 1.0 {
                    filters.push(format!("eq=saturation={amount:.3}"));
                }
            }
            "blur" => {
                // CSS blur() takes a radius; ffmpeg's gblur takes a standard
                // deviation, which is roughly half of it.
                let radius = effect.param("radius", 0.0);
                if radius > 0.0 {
                    filters.push(format!("gblur=sigma={:.2}", radius / 2.0));
                }
            }
            "hue" => {
                let angle = effect.param("angle", 0.0);
                if angle != 0.0 {
                    filters.push(format!("hue=h={}", angle.round() as i64));
                }
            }
            "grayscale" => {
                let amount = effect.param("amount", 0.0);
                if amount > 0.0 {
                    filters.push(format!("hue=s={:.3}", 1.0 - amount));
                }
            }
            "invert" => {
                let amount = effect.param("amount", 0.0);
                if amount > 0.0 {
                    filters.push(invert_filter(amount));
                }
            }
            "rgbsplit" => {
                if let Some(filter) =
                    rgb_split_filter(effect.param("amount", 0.0), effect.param("angle", 0.0))
                {
                    filters.push(filter);
                }
            }
            "motionblur" => {
                if let Some(filter) =
                    motion_blur_filter(effect.param("amount", 0.0), effect.param("angle", 0.0))
                {
                    filters.push(filter);
                }
            }
            unknown => {
                // Forward-compatible: a document written by a newer UI keeps
                // its effect, and the encoder simply ignores what it cannot map.
                eprintln!("veglass: effet inconnu ignoré au rendu : {unknown}");
            }
        }
    }

    filters
}

pub fn describe(effects: &[Effect]) -> EffectChain {
    let filters = effect_filters(effects);
    EffectChain {
        chain: filters.join(","),
        filters,
        engine: "rust".to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    fn effect(kind: &str, key: &str, value: f64) -> Effect {
        let mut params = HashMap::new();
        params.insert(key.to_string(), value);
        Effect { id: "fx".into(), kind: kind.into(), enabled: true, params }
    }

    /// These strings are the contract with `src/lib/effectChain.ts`; the twin
    /// must produce them character for character.
    #[test]
    fn maps_each_effect_kind() {
        assert_eq!(effect_filters(&[effect("brightness", "amount", 0.15)]), ["eq=brightness=0.150"]);
        assert_eq!(effect_filters(&[effect("contrast", "amount", 1.2)]), ["eq=contrast=1.200"]);
        assert_eq!(effect_filters(&[effect("saturation", "amount", 1.35)]), ["eq=saturation=1.350"]);
        assert_eq!(effect_filters(&[effect("blur", "radius", 6.0)]), ["gblur=sigma=3.00"]);
        assert_eq!(effect_filters(&[effect("hue", "angle", 30.0)]), ["hue=h=30"]);
        assert_eq!(effect_filters(&[effect("grayscale", "amount", 1.0)]), ["hue=s=0.000"]);
        assert_eq!(
            effect_filters(&[effect("invert", "amount", 1.0)]),
            ["lutrgb=r=val*(-1.000)+(1.000)*maxval:g=val*(-1.000)+(1.000)*maxval:b=val*(-1.000)+(1.000)*maxval"]
        );
        assert_eq!(
            effect_filters(&[effect("rgbsplit", "amount", 10.0)]),
            ["rgbashift=rh=10:rv=0:bh=-10:bv=0:edge=smear"]
        );
        assert_eq!(
            effect_filters(&[effect("motionblur", "amount", 14.0)]),
            ["gblur=sigma=14.00:sigmaV=0.00:steps=2"]
        );
    }

    /// The directions were read off the binary, and this pins what was read:
    /// red right and blue left at 0°, red down and blue up at 90°. SVG's
    /// `feOffset` uses the same convention, which is what makes the viewer and
    /// the file show the same fringe on the same side.
    #[test]
    fn a_split_moves_red_and_blue_opposite_ways() {
        assert_eq!(
            rgb_split_filter(10.0, 0.0).unwrap(),
            "rgbashift=rh=10:rv=0:bh=-10:bv=0:edge=smear"
        );
        assert_eq!(
            rgb_split_filter(8.0, 90.0).unwrap(),
            "rgbashift=rh=0:rv=8:bh=0:bv=-8:edge=smear"
        );
        // Under half a pixel there is nothing to shift, and a filter that moves
        // nothing is a pass spent on nothing.
        assert!(rgb_split_filter(0.4, 0.0).is_none());
        assert!(rgb_split_filter(0.0, 45.0).is_none());
    }

    /// A blur has an axis, not a direction: 0° and 180° are the same blur, and
    /// the deviation is split across the axes rather than rotated.
    #[test]
    fn a_motion_blur_stays_on_one_axis() {
        assert_eq!(
            motion_blur_filter(14.0, 0.0).unwrap(),
            "gblur=sigma=14.00:sigmaV=0.00:steps=2"
        );
        assert_eq!(
            motion_blur_filter(14.0, 180.0).unwrap(),
            motion_blur_filter(14.0, 0.0).unwrap()
        );
        assert_eq!(
            motion_blur_filter(14.0, 90.0).unwrap(),
            "gblur=sigma=0.00:sigmaV=14.00:steps=2"
        );
        assert!(motion_blur_filter(0.0, 0.0).is_none());
    }

    /// The two ends of the dosage, checked as arithmetic rather than as strings.
    #[test]
    fn a_full_invert_is_the_negative_and_a_half_one_is_flat_grey() {
        // out = in·(1−2a) + a·max, the line CSS `invert()` defines.
        let at = |amount: f64, input: f64| input * (1.0 - 2.0 * amount) + amount;

        assert!((at(1.0, 0.0) - 1.0).abs() < 1e-9, "black inverts to white");
        assert!((at(1.0, 1.0) - 0.0).abs() < 1e-9, "white inverts to black");
        assert!((at(0.5, 0.0) - 0.5).abs() < 1e-9, "half-way is mid-grey whatever went in");
        assert!((at(0.5, 1.0) - 0.5).abs() < 1e-9);
        assert!((at(0.0, 0.7) - 0.7).abs() < 1e-9, "no dosage changes nothing");
    }

    #[test]
    fn skips_neutral_and_disabled_effects() {
        assert!(effect_filters(&[effect("brightness", "amount", 0.0)]).is_empty());
        assert!(effect_filters(&[effect("contrast", "amount", 1.0)]).is_empty());

        let mut off = effect("blur", "radius", 8.0);
        off.enabled = false;
        assert!(effect_filters(&[off]).is_empty());
    }

    #[test]
    fn preserves_stack_order_in_the_chain() {
        let chain = describe(&[
            effect("blur", "radius", 4.0),
            effect("saturation", "amount", 2.0),
        ]);
        assert_eq!(chain.chain, "gblur=sigma=2.00,eq=saturation=2.000");
        assert_eq!(chain.engine, "rust");
    }

    #[test]
    fn unknown_kinds_are_ignored_not_fatal() {
        assert!(effect_filters(&[effect("bokeh-from-a-newer-build", "amount", 1.0)]).is_empty());
    }
}
